libs/langchain/langchain/chains/conversational_retrieval/base.py

```diff
@@ -465,3 +467,59 @@ def from_llm(
             callbacks=callbacks,
             **kwargs,
         )
+
+
+def create_chat_retriever_chain(
+    llm: LanguageModelLike,
+    retriever: RetrieverLike,
+    prompt: BasePromptTemplate,
+) -> RetrieverOutputLike:
+    """Create a chain that takes conversation history and returns documents.
+
+    If there is no `chat_history`, then the `input` is just passed directly to the
+    retriever. If there is `chat_history`, then the prompt and LLM will be used
+    to generate a search query. That search query is then passed to the retriever.
+
+    Args:
+        llm: Language model to use for generating a search term given chat history
+        retriever: RetrieverLike object that takes a string as input and outputs
+            a list of Documents.
+        prompt: The prompt used to generate the search query for the retriever.
+
+    Returns:
+        An LCEL Runnable. The runnable input must take in `input`, and if there
+        is chat history should take it in the form of `chat_history`.
+        The Runnable output is a list of Documents
+
+    Example:
+        .. code-block:: python
+
+            # pip install -U langchain langchain-community
+
+            from langchain_community.chat_models import ChatOpenAI
+            from langchain.chains import create_chat_retriever_chain
+            from langchain import hub
+
+            rephrase_prompt = hub.pull("langchain-ai/chat-langchain-rephrase")
+            llm = ChatOpenAI()
+            retriever = ...
+            chat_retriever_chain = create_chat_retriever_chain(
+                llm, retriever, rephrase_prompt
+            )
+
+            chain.invoke({"input": "...", "chat_history": })
+
+    """
+    if "input" not in prompt.input_variables:
+        raise ValueError(
+            "Expected `input` to be a prompt variable, "
+            f"but got {prompt.input_variables}"
+        )
+    retrieve_documents = RunnableBranch(
+        (
+            lambda x: len(x.get("chat_history", [])) == 0,
+            (lambda x: x["input"]) | retriever,
+        ),
+        prompt | llm | StrOutputParser() | retriever,
```

Comment:
So `chat_history` would have to be strictly a string or a list of messages depending on the prompt right?

Reply:
yeah

people can always add to start if they way?

Reply:
Yeah not opposed, just want to make sure